core/client/fs/src/main/java/alluxio/client/file/cache/store/LocalPageStore.java

```diff
@@ -61,43 +72,76 @@ public LocalPageStore(LocalPageStoreOptions options) {
     mFileBuckets = options.getFileBuckets();
     // normalize the path to deal with trailing slash
     Path rootDir = Paths.get(mRoot);
-    // pattern encoding root_path/page_size(ulong)/bucket(uint)/file_id(str)/page_idx(ulong)/
-    mPagePattern = Pattern.compile(
-        String.format("%s/%d/(\\d+)/([^/]+)/(\\d+)", Pattern.quote(rootDir.toString()), mPageSize));
+    // pattern encoding for each page
+    // root_path/page_size(ulong)/bucket(uint)/file_id(str)/mtime(ulong)/page_idx(ulong)/
+    mPagePattern = Pattern.compile(String
+        .format("%s/%d/(\\d+)/([^/]+)/(\\d+)/(\\d+)", Pattern.quote(rootDir.toString()),
+            mPageSize));
+    // pattern encoding for the directory of each file
+    // root_path/page_size(ulong)/bucket(uint)/file_id(str)/
+    mFilePattern = Pattern.compile(
+        String.format("%s/%d/(\\d+)/([^/]+)", Pattern.quote(rootDir.toString()), mPageSize));
   }
 
   @Override
-  public void put(PageId pageId, byte[] page) throws ResourceExhaustedException, IOException {
-    Path p = getFilePath(pageId);
+  public void put(PageInfo pageInfo, byte[] page)
+      throws ResourceExhaustedException, IOException {
+    PageId pageId = pageInfo.getPageId();
+    Path pagePath = getPageFilePath(pageInfo);
+    Path timestampPath = Preconditions
+        .checkNotNull(pagePath.getParent(), "timestamp path should not be null");
     try {
-      if (!Files.exists(p)) {
-        Path parent =
-            Preconditions.checkNotNull(p.getParent(), "parent of cache file should not be null");
-        Files.createDirectories(parent);
-        Files.createFile(p);
+      if (!Files.exists(pagePath)) {
+        if (!Files.exists(timestampPath)) { //new timestamp, new file or file got updated
+          Path fileLevelPath = Preconditions
+              .checkNotNull(timestampPath.getParent(), "file level path should not be null");
+          if (Files.exists(fileLevelPath)) {
+            //clear stale timestamp folders
+            FileUtils.cleanDirectory(fileLevelPath.toFile());
```

Comment:
Race condition will throw FileNotFound exception caused by another thread delete the file.

```
2022-02-18 12:36:53,108 [pool-8-thread-3] ERROR cache.LocalCacheManager (LocalCacheManager.java:putAttempt) - Failed to add page PageId{FileId=5, PageIndex=2} to pageStore
java.io.IOException: Failed to write file /var/folders/hj/jty18sr93t3_cbkb01szbv3w0000gn/T/junit2422623433926909480/LOCAL/1024/53/5/0 for page PageId{FileId=5, PageIndex=2}
	at alluxio.client.file.cache.store.LocalPageStore.put(LocalPageStore.java:126)
	at alluxio.client.file.cache.HangingPageStore.put(HangingPageStore.java:60)
	at alluxio.client.file.cache.LocalCacheManager.putAttempt(LocalCacheManager.java:384)
	at alluxio.client.file.cache.LocalCacheManager.putInternal(LocalCacheManager.java:303)
	at alluxio.client.file.cache.LocalCacheManager.lambda$put$1(LocalCacheManager.java:278)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.IOException: Unable to delete file: /var/folders/hj/jty18sr93t3_cbkb01szbv3w0000gn/T/junit2422623433926909480/LOCAL/1024/53/5/0
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1338)
	at org.apache.commons.io.FileUtils.cleanDirectory(FileUtils.java:334)
	at alluxio.client.file.cache.store.LocalPageStore.put(LocalPageStore.java:104)
	... 10 more
Caused by: java.nio.file.NoSuchFileException: /var/folders/hj/jty18sr93t3_cbkb01szbv3w0000gn/T/junit2422623433926909480/LOCAL/1024/53/5/0/11
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1763)
	at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:219)
	at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:276)
	at java.base/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:373)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2760)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2796)
	at org.apache.commons.io.file.PathUtils.visitFileTree(PathUtils.java:535)
	at org.apache.commons.io.file.PathUtils.deleteDirectory(PathUtils.java:265)
	at org.apache.commons.io.file.PathUtils.delete(PathUtils.java:254)
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1336)
	... 12 more
```

Reply:
@beinan I suggest to change the `getPageLockId` to `getPageFileLockId` like the following.

```
  private int getPageLockId(PageId pageId) {
    return Math.floorMod((pageId.getFileId().hashCode()), LOCK_SIZE);
  }
```